horovod/run/common/service/driver_service.py

```diff
@@ -95,13 +107,25 @@ def _filter_by_ip(self, addresses, target_ip):
         return {}
 
     def all_task_addresses(self, index):
-        return self._all_task_addresses[index].copy()
+        self._wait_cond.acquire()
```

Comment:
Have to use the `_wait_cond` here because we want to read these variables while they are not being modified in handling `RegisterTask*` requests. A different lock would not prevent that concurrency.